### ISSUE TITLE
Historical metrics being written to StatsD inconsistently

### ISSUE THREAD

**Issue by ryanrolds** [mperham/sidekiq#4873]
Ruby version: 2.5.1
Rails version: 5.2.2
Sidekiq / Pro / Enterprise version(s): 5.2.5, 4.0.5, 1.8.1

Part of the initializer:
```
...

  Sidekiq::Pro.dogstatsd = ->{
    Datadog::Statsd.new(
      Rails.application.config.statsd_host,
      Rails.application.config.statsd_port,
      tags: {
        service: Rails.application.class.parent_name.downcase,
        environment: Rails.env.downcase,
      }
    )
  }

  Sidekiq.configure_server do |config|
    if Rails.application.config.redis_url
      config.redis = { url: Rails.application.config.redis_url }
    end

    # Enabling enterprise history
    # https://github.com/mperham/sidekiq/wiki/Ent-Historical-Metrics
    config.retain_history(30)

    # Send job stats to StatsD
    config.server_middleware do |chain|
      require 'sidekiq/middleware/server/statsd'
      chain.add Sidekiq::Middleware::Server::Statsd
    end
    
    config.periodic do |mgr|
    
      ...

      # For debugging purposes, fire off jobs every other minute
      if Rails.env.development?
        mgr.register('*/2 * * * *', NoOpJob)
        mgr.register('*/2 * * * *', NoOpJobLow)
      end
    end
  end
```

I've got StatsD in debug mode and I'm watching metrics come in. I've got two jobs that run every 2 minutes that do nothing. On startup I'm seeing the historical metrics and I'm seeing the "pro metrics" for the jobs that run every other minute. After the initial emitting of historical metrics I no longer see them being emitted. If I adjust the jobs to run every 1 minute, then I do see the historical metrics being emitted. However the metrics are only emitted every minute, not every 30 seconds as I expect.

My expectation is that historical metrics would be emitted every 30 seconds, even if there were no jobs processed.

**Comment by mperham** [comment]
One tip: it’s a bad idea to create one global instance: dd_statsd. Put that new within the dogstatsd Proc as documented so Sidekiq can use a pool of instances.

**Comment by ryanrolds** [comment]
Will that tip resolve the issue with StatsD metric not being emitted or is it unrelated?

**Comment by mperham** [comment]
This definitely could be the cause but I can't say for sure.

**Comment by ryanrolds** [comment]
I've adjusted the initializer per my understanding of your suggestion. The behavior is the same. 

StatsD debug output: https://gist.github.com/ryanrolds/062da4df6c29085a853c3c713a8fd8de

**Comment by mperham** [comment]
The other obviously wrong thing in your initializer is `*\2 * * * *`. To run every other minute, you use slash, not backslash: `*/2 * * * *`. Maybe that's why you aren't seeing the job metrics on the expected schedule?

**Comment by ryanrolds** [comment]
I've switched it. Updated the post at the top. No change in behavior. The jobs were running at the right interval, then and now.

**Comment by mperham** [comment]
I'm kinda stumped. Can you start Sidekiq with `-v`, run it for a few minutes and post the log output?

**Comment by ryanrolds** [comment]
The only time historical metrics where output was right after leader election completed during startup.

https://gist.github.com/ryanrolds/165af4de34a5d0b3af44b6be17ffd442

**Comment by ryanrolds** [comment]
I've added some logger calls in `lib/sidekiq-ent/metrics.rb` and have confirmed that "run" thread is ticking every 30 seconds and `Sidekiq::Senate.leader?` is true. I'm digging down deeper now.

**Comment by mperham** [comment]
The banner prints if the output is a TTY. Can you turn off your custom logger and use the default logger so I can see the Thread ID associated with each line?

**Comment by ryanrolds** [comment]
https://gist.github.com/ryanrolds/e2ee269fa919d1cf459ab363423c504b

**Comment by mperham** [comment]
One suspicious behavior is this double Tick, at :59 and :00. I've never seen that before.

```
2021-04-14T20:18:00.014Z 475 TID-gpnxvf9yn DEBUG: Tick: [2021-04-14 20:18:00 +0000, "d06af3f77e05d58fea3f15604d659ec4e0ec3d1c", "178930f631332ce58e8cfb141edb3de077b65dd5"]
...
2021-04-14T20:18:59.934Z 475 TID-gpnxvf9yn DEBUG: Tick: [2021-04-14 20:18:59 +0000]
2021-04-14T20:19:00.001Z 475 TID-gpnxvf9yn DEBUG: Tick: [2021-04-14 20:19:00 +0000]
2021-04-14T20:19:59.934Z 475 TID-gpnxvf9yn DEBUG: Tick: [2021-04-14 20:19:59 +0000]
```

But I can't reproduce this. v1.8.1 ticks consistently at :00 for me with Rails 5.2.2. Can you supply a tiny Rails app which reproduces the problem?

**Comment by ryanrolds** [comment]
TL;DR Looks like it was UDP MTU issues.

I think I figured it out. Job-level metrics are not too large to batch, but the historical metrics can be too large. I traced the call path down into dogstatsd-ruby and confirmed that the metrics are being sent every 30 seconds and are being written out on the UDP socket. I'm not sure why the initial batch didn't file, it may be doing a different path initially.

Once I set `max_buffer_bytes` to 1500 on the DataDog StatsD client the problem goes away. It looks like it's default max is 8192.
